In [21]:
!pip3 install openpyxl streamlit pandas plotly



[notice] A new release of pip is available: 24.1.2 -> 24.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
import pandas as pd
import plotly.express as px

# Pricing Dataset

In [4]:
pricing_df = pd.read_csv("src/get_around_pricing_project.csv", index_col=[0])

In [5]:
pricing_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4843 entries, 0 to 4842
Data columns (total 14 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   model_key                  4843 non-null   object
 1   mileage                    4843 non-null   int64 
 2   engine_power               4843 non-null   int64 
 3   fuel                       4843 non-null   object
 4   paint_color                4843 non-null   object
 5   car_type                   4843 non-null   object
 6   private_parking_available  4843 non-null   bool  
 7   has_gps                    4843 non-null   bool  
 8   has_air_conditioning       4843 non-null   bool  
 9   automatic_car              4843 non-null   bool  
 10  has_getaround_connect      4843 non-null   bool  
 11  has_speed_regulator        4843 non-null   bool  
 12  winter_tires               4843 non-null   bool  
 13  rental_price_per_day       4843 non-null   int64 
dtypes: bool(7), i

In [6]:
pricing_df.describe(include="all")

model_key       mileage  engine_power    fuel paint_color car_type  \
count       4843  4.843000e+03    4843.00000    4843        4843     4843   
unique        28           NaN           NaN       4          10        8   
top      Citroën           NaN           NaN  diesel       black   estate   
freq         969           NaN           NaN    4641        1633     1606   
mean         NaN  1.409628e+05     128.98823     NaN         NaN      NaN   
std          NaN  6.019674e+04      38.99336     NaN         NaN      NaN   
min          NaN -6.400000e+01       0.00000     NaN         NaN      NaN   
25%          NaN  1.029135e+05     100.00000     NaN         NaN      NaN   
50%          NaN  1.410800e+05     120.00000     NaN         NaN      NaN   
75%          NaN  1.751955e+05     135.00000     NaN         NaN      NaN   
max          NaN  1.000376e+06     423.00000     NaN         NaN      NaN   

       private_parking_available has_gps has_air_conditioning automatic_car  \
count                       4843    4843                 4843          4843   
unique                         2       2                    2             2   
top                         True    True                False         False   
freq                        2662    3839                 3865          3881   
mean                         NaN     NaN                  NaN           NaN   
std                          NaN     NaN                  NaN           NaN   
min                          NaN     NaN                  NaN           NaN   
25%                          NaN     NaN                  NaN           NaN   
50%                          NaN     NaN                  NaN           NaN   
75%                          NaN     NaN                  NaN           NaN   
max                          NaN     NaN                  NaN           NaN   

       has_getaround_connect has_speed_regulator winter_tires  \
count                   4843                4843         4843   
unique                     2                   2            2   
top                    False               False         True   
freq                    2613                3674         4514   
mean                     NaN                 NaN          NaN   
std                      NaN                 NaN          NaN   
min                      NaN                 NaN          NaN   
25%                      NaN                 NaN          NaN   
50%                      NaN                 NaN          NaN   
75%                      NaN                 NaN          NaN   
max                      NaN                 NaN          NaN   

        rental_price_per_day  
count            4843.000000  
unique                   NaN  
top                      NaN  
freq                     NaN  
mean              121.214536  
std                33.568268  
min                10.000000  
25%               104.000000  
50%               119.000000  
75%               136.000000  
max               422.000000

In [7]:
display(pricing_df.describe())

mileage  engine_power  rental_price_per_day
count  4.843000e+03    4843.00000           4843.000000
mean   1.409628e+05     128.98823            121.214536
std    6.019674e+04      38.99336             33.568268
min   -6.400000e+01       0.00000             10.000000
25%    1.029135e+05     100.00000            104.000000
50%    1.410800e+05     120.00000            119.000000
75%    1.751955e+05     135.00000            136.000000
max    1.000376e+06     423.00000            422.000000

In [8]:
display(pricing_df.head())

model_key  mileage  engine_power    fuel paint_color     car_type  \
0   Citroën   140411           100  diesel       black  convertible   
1   Citroën    13929           317  petrol        grey  convertible   
2   Citroën   183297           120  diesel       white  convertible   
3   Citroën   128035           135  diesel         red  convertible   
4   Citroën    97097           160  diesel      silver  convertible   

   private_parking_available  has_gps  has_air_conditioning  automatic_car  \
0                       True     True                 False          False   
1                       True     True                 False          False   
2                      False    False                 False          False   
3                       True     True                 False          False   
4                       True     True                 False          False   

   has_getaround_connect  has_speed_regulator  winter_tires  \
0                   True                 True          True   
1                  False                 True          True   
2                   True                False          True   
3                   True                 True          True   
4                  False                 True          True   

   rental_price_per_day  
0                   106  
1                   264  
2                   101  
3                   158  
4                   183

In [9]:
# Removing extreme outliers in mileage and price
pricing_cleaned = pricing_df[
    (pricing_df['mileage'] > 0) & 
    (pricing_df['rental_price_per_day'] > 0)
]

price_dist_fig = px.histogram(
    pricing_cleaned, 
    x='rental_price_per_day', 
    title='Distribution of Rental Prices',
    nbins=30, 
    labels={'rental_price_per_day': 'Rental Price per Day'},
    color_discrete_sequence=['#636EFA']
)

mileage_price_fig = px.scatter(
    pricing_cleaned, 
    x='mileage', 
    y='rental_price_per_day', 
    title='Rental Price vs Mileage',
    labels={'mileage': 'Mileage', 'rental_price_per_day': 'Rental Price per Day'},
    color_discrete_sequence=['#00CC96']
)

price_dist_fig.show()
mileage_price_fig.show()

### Conclusions

* Mileage: ~140,962 km (data between 64km and 1,000,376km)
* Engine Power: ~129 HP
* Rental Price per Day: ~121.2 units (data between 10 and 422 units)

# Delay Dataset

In [10]:
delay_df = pd.read_excel("src/get_around_delay_analysis.xlsx")

In [11]:
display(delay_df.head())

rental_id  car_id checkin_type     state  delay_at_checkout_in_minutes  \
0     505000  363965       mobile  canceled                           NaN   
1     507750  269550       mobile     ended                         -81.0   
2     508131  359049      connect     ended                          70.0   
3     508865  299063      connect  canceled                           NaN   
4     511440  313932       mobile     ended                           NaN   

   previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
0                       NaN                                         NaN  
1                       NaN                                         NaN  
2                       NaN                                         NaN  
3                       NaN                                         NaN  
4                       NaN                                         NaN

In [12]:
delay_df["state"].value_counts()

state
ended       18045
canceled     3265
Name: count, dtype: int64

In [13]:
# Filter ended rentals and summarize delay statistics
ended_rentals = delay_df[delay_df["state"] == "ended"]
display(ended_rentals["delay_at_checkout_in_minutes"].describe())

# Count rentals with delays and without delays
display(ended_rentals["delay_at_checkout_in_minutes"].notnull().value_counts())


count    16345.000000
mean        60.773876
std        993.173222
min     -22433.000000
25%        -36.000000
50%          9.000000
75%         67.000000
max      71084.000000
Name: delay_at_checkout_in_minutes, dtype: float64

delay_at_checkout_in_minutes
True     16345
False     1700
Name: count, dtype: int64

In [14]:
delay_cleaned = ended_rentals[
    (ended_rentals['delay_at_checkout_in_minutes'] > -5000) & 
    (ended_rentals['delay_at_checkout_in_minutes'] < 5000)
]

delay_dist_fig = px.histogram(
    delay_cleaned, 
    x='delay_at_checkout_in_minutes', 
    title='Distribution of Checkout Delays',
    nbins=50, 
    labels={'delay_at_checkout_in_minutes': 'Delay at Checkout (Minutes)'}
)

delay_dist_fig.show()

In [16]:
# Define thresholds for minimum delay in minutes (1 hour, 2 hours, 3 hours)
thresholds = [60, 120, 180]

# Calculate number of rentals affected for each threshold and type of check-in
affected_rentals = []
for threshold in thresholds:
    affected = delay_cleaned[
        (delay_cleaned['time_delta_with_previous_rental_in_minutes'] < threshold) &
        (delay_cleaned['time_delta_with_previous_rental_in_minutes'] > 0)
    ]
    affected_counts = affected['checkin_type'].value_counts().reset_index()
    affected_counts.columns = ['checkin_type', 'count']
    affected_counts['threshold'] = threshold
    affected_rentals.append(affected_counts)

# Concatenate all affected counts into a single DataFrame
affected_rentals_df = pd.concat(affected_rentals, ignore_index=True)

# Visualization: Number of rentals affected by threshold and check-in type
affected_rentals_fig = px.bar(
    affected_rentals_df,
    x='threshold',
    y='count',
    color='checkin_type',
    barmode='group',
    title='Number of Rentals Affected by Minimum Delay Threshold and Check-In Type',
    labels={
        'threshold': 'Minimum Delay Threshold (Minutes)',
        'count': 'Number of Rentals Affected',
        'checkin_type': 'Check-In Type'
    },
    color_discrete_sequence=px.colors.qualitative.Set1
)

# Show the visualization
affected_rentals_fig.show()


### Conclusions

* Mean delay ~60.7 minutes whereas median delay ~9 minutes.
* Minimum and maximum values may be outliers or errors.

In [20]:
# Filter positive delays (only actual delays)
positive_delays = delay_cleaned[delay_cleaned['delay_at_checkout_in_minutes'] > 0]

# Define thresholds to analyze (in minutes)
delay_thresholds = [30, 60, 120]

# Count the number of delays greater than each threshold
delay_counts = []
for threshold in delay_thresholds:
    count = positive_delays[positive_delays['delay_at_checkout_in_minutes'] > threshold].shape[0]
    delay_counts.append({'threshold': threshold, 'count': count})

# Convert results to DataFrame
delay_counts_df = pd.DataFrame(delay_counts)

# Visualization: Number of rentals delayed beyond thresholds
delay_threshold_fig = px.bar(
    delay_counts_df,
    x='threshold',
    y='count',
    title='Number of Rentals Delayed Beyond Thresholds',
    labels={
        'threshold': 'Delay Threshold (Minutes)',
        'count': 'Number of Rentals'
    }
)

# Show the visualization
delay_threshold_fig.show()
